File: library/src/main/java/im/toss/delegationadapter/ListItemAdapterDelegate.java
package im.toss.delegationadapter;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.ViewGroup;

import com.hannesdorfmann.adapterdelegates3.AbsListItemAdapterDelegate;

import java.util.List;

/**
 * More simple way to create {@link AbsListItemAdapterDelegate} with less boiler plate code
 * @param <T> The type of the item that is managed by this AdapterDelegate.
 */
public class ListItemAdapterDelegate<I extends T, T> extends AbsListItemAdapterDelegate<I, T, ItemViewHolder> {

    @LayoutRes
    private int layoutId;
    private ViewTypeChecker<T> typeChecker;
    private ViewHolderCreateListener onViewHolderCreate;
    private ViewBinder<I> viewBinder;
    private ViewBinderWithoutPayloads<I> viewBinderWithoutPayloads;

    private ListItemAdapterDelegate(Builder<I,T> builder) {
        layoutId = builder.layoutId;
        typeChecker = builder.typeChecker;
        onViewHolderCreate = builder.onViewHolderCreate;
        viewBinder = builder.viewBinder;
        viewBinderWithoutPayloads = builder.viewBinderWithoutPayloads;
    }

    @Override
    protected boolean isForViewType(@NonNull T item, @NonNull List<T> items, int position) {
        if (typeChecker != null) {
            return typeChecker.isForViewType(item);
        }
        return false;
    }

    @NonNull
    @Override
    protected ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent) {
        ItemViewHolder viewHolder = ItemViewHolder.create(parent, layoutId);
        if (onViewHolderCreate != null) {
            onViewHolderCreate.onViewHolderCreate(viewHolder);
        }
        return viewHolder;
    }

    @Override
    protected void onBindViewHolder(@NonNull I item, @NonNull ItemViewHolder viewHolder, @NonNull List<Object> payloads) {
        if (viewBinder != null) {
            viewBinder.onBindViewHolder(viewHolder, item, payloads);
        } else if (viewBinderWithoutPayloads != null) {
            viewBinderWithoutPayloads.onBindViewHolder(viewHolder, item);
        }
    }

    public static class Builder<I extends T, T> {
        private int layoutId;
        private ViewTypeChecker<T> typeChecker;
        private ViewHolderCreateListener onViewHolderCreate;
        private ViewBinder<I> viewBinder;
        private ViewBinderWithoutPayloads<I> viewBinderWithoutPayloads;

        Builder() {
        }

        Builder<I,T> typeChecker(ViewTypeChecker<T> val) {
            typeChecker = val;
            return this;
        }

        public Builder<I,T> layout(@LayoutRes int layoutId) {
            this.layoutId = layoutId;
            return this;
        }

        public Builder<I,T> binder(ViewBinder<I> val) {
            viewBinder = val;
            return this;
        }

        public Builder<I,T> binder(ViewBinderWithoutPayloads<I> val) {
            viewBinderWithoutPayloads = val;
            return this;
        }

        public Builder<I,T> onViewHolderCreate(ViewHolderCreateListener callback) {
            this.onViewHolderCreate = callback;
            return this;
        }

        public ListItemAdapterDelegate<I,T> build() {
            return new ListItemAdapterDelegate<>(this);
        }
    }
}

File: library-kotlin/src/main/kotlin/im/toss/delegationadapter/kotlin/ListItemAdapterDelegate.kt
package im.toss.delegationadapter.kotlin

import android.support.annotation.LayoutRes
import android.view.ViewGroup
import com.hannesdorfmann.adapterdelegates3.AbsListItemAdapterDelegate


/**
 * More simple way to create [AbsListItemAdapterDelegate] with less boiler plate code
 * @param <T> The type of the item that is managed by this AdapterDelegate.
</T> */
class ListItemAdapterDelegate<I : T, T> constructor(
    @LayoutRes val layoutId: Int,
    val typeChecker: ((T) -> Boolean)? = null,
    val onViewHolderCreate: ViewHolderCreateListener? = null,
    val viewBinder: ((ItemViewHolder<I>, I) -> Unit)? = null,
    val viewBinderLong: ((ItemViewHolder<I>, I, List<Any>?) -> Unit)? = null
) : AbsListItemAdapterDelegate<I, T, ItemViewHolder<I>>() {

  override fun isForViewType(item: T, items: List<T>, position: Int): Boolean {
    return typeChecker?.invoke(item) ?: false
  }

  override fun onCreateViewHolder(parent: ViewGroup): ItemViewHolder<I> {
    val viewHolder = ItemViewHolder.create<I>(parent, layoutId)
    onViewHolderCreate?.onViewHolderCreate(viewHolder)
    return viewHolder
  }

  override fun onBindViewHolder(item: I, viewHolder: ItemViewHolder<I>, payloads: List<Any>) {
    viewHolder.item = item
    viewBinderLong?.invoke(viewHolder, item, payloads) ?:
        viewBinder?.invoke(viewHolder, item)
  }


  class Builder<I : T, T> {
    private var layoutId: Int = 0
    private var typeChecker: ((T) -> Boolean)? = null
    private var onViewHolderCreate: ViewHolderCreateListener? = null
    private var viewBinder: ((ItemViewHolder<I>, I) -> Unit)? = null
    private var viewBinderLong: ((ItemViewHolder<I>, I, List<Any>?) -> Unit)? = null

    fun typeChecker(typeChecker: (T) -> Boolean): Builder<I, T> {
      this.typeChecker = typeChecker
      return this
    }

    fun layout(@LayoutRes layoutId: Int): Builder<I, T> {
      this.layoutId = layoutId
      return this
    }

    fun binder(binder: (ItemViewHolder<I>, I) -> Unit): Builder<I, T> {
      this.viewBinder = binder
      return this
    }

    fun binder(binder: (ItemViewHolder<I>, I, List<Any>?) -> Unit): Builder<I, T> {
      this.viewBinderLong = binder
      return this
    }

    fun onViewHolderCreate(callback: ViewHolderCreateListener): Builder<I, T> {
      this.onViewHolderCreate = callback
      return this
    }

    fun build(): ListItemAdapterDelegate<I, T> {
      return ListItemAdapterDelegate(
          layoutId = layoutId,
          typeChecker = typeChecker,
          viewBinder = viewBinder,
          viewBinderLong = viewBinderLong,
          onViewHolderCreate = onViewHolderCreate
      )
    }
  }
}
